src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Completion.cs

```diff
@@ -271,4 +282,135 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
+    public CompletionItem[]? GetSnippetCompletions(RazorSnippetCompletionParams completionParams, CancellationToken _)
+    {
+        if (completionParams.TextSpan.Length == 0)
+        {
+            return null;
+        }
+
+        var snippets = _snippetCache.GetSnippets();
+
+        if (snippets.IsDefaultOrEmpty)
+        {
+            return null;
+        }
+
+        var langSpecificCompletionsEnum = completionParams.LanguageKind switch
+        {
+            RazorLanguageKind.Html => snippets.Where(static s => s.Language == SnippetLanguage.Html),
+            RazorLanguageKind.CSharp => snippets.Where(static s => s.Language == SnippetLanguage.CSharp),
+            _ => null
+        };
+
+        if (langSpecificCompletionsEnum is null)
+        {
+            return null;
+        }
+
+        if (!_documentManager.TryGetDocument(completionParams.Identifier.Uri, out var snapshot))
+        {
+            return null;
+        }
+
+        var text = snapshot.Snapshot.GetText(completionParams.TextSpan.Start, completionParams.TextSpan.Length);
+
+        var completions = langSpecificCompletionsEnum
+            .Where(s => s.Shortcut.StartsWith(text, StringComparison.OrdinalIgnoreCase))
+            .Select(s => (s, _xmlSnippetParser.GetParsedXmlSnippet(s)))
+            .Select(ConvertToCompletionItem)
+            .WithoutNull();
+
+        return completions.ToArray();
+
+        CompletionItem? ConvertToCompletionItem((SnippetInfo info, ParsedXmlSnippet? parsed) pair)
+        {
+            var (info, parsed) = pair;
+
+            if (info is null || parsed is null)
+            {
+                return null;
+            }
+
+            return new()
+            {
+                Label = info.Shortcut,
+                Detail = info.Description,
+                InsertTextFormat = InsertTextFormat.Snippet,
+                InsertText = GetFormattedLspSnippet(info, parsed, completionParams.TextSpan)
+            };
+        }
+    }
+
+    /// <summary>
+    /// Formats the snippet by applying the snippet to the document with the default values / function results for snippet declarations.
+    /// Then converts back into an LSP snippet by replacing the declarations with the appropriate LSP tab stops.
+    /// 
+    /// Note that the operations in this method are sensitive to the context in the document and so must be calculated on each request.
+    /// </summary>
+    private static string GetFormattedLspSnippet(
+        SnippetInfo snippetInfo,
+        ParsedXmlSnippet parsedSnippet,
+        RazorTextSpan snippetShortcut)
+    {
+        // Calculate the snippet text with defaults + snippet function results.
+        var (snippetFullText, fields, caretSpan) = GetReplacedSnippetText(snippetInfo, snippetShortcut, parsedSnippet);
+
+        // TODO: use the fields? I believe those end up being tab stops
+        return snippetFullText.ToString();
```

Comment:
`snippetFullText` is already a string, right? Or is this changing with the above TODO?